CNDB-10463: Replace all usages of PriorityQueue in SAI with efficient versions

blambov opened the pull request.

pkolaczk commented on test/unit/org/apache/cassandra/utils/SortingIteratorTest.java:
```diff
@@ -0,0 +1,417 @@
+
+        iterator.skipTo(4);
+        assertTrue(iterator.hasNext());
+        assertEquals((Integer) 4, iterator.next());
```
suggestion: Should check if it skipped to the first 4. So check if we get the 4 three times.

blambov replied:
Fixed

pkolaczk commented on src/java/org/apache/cassandra/utils/SortingIterator.java:
```diff
@@ -0,0 +1,317 @@
+ * steps. skipTo works in O(m.log n) steps (where m is the number of skipped items), but is also limited to O(n) when m
+ * is large by the same argument as the initial heap construction.
+ * <p>
+ * The accepts and stores nulls as non-present values, which turns out to be quite a bit more efficient for iterating
```
Something wrong with the grammar here at the beginning of this sentence. "The <what> accepts and stores nulls ..." ?

blambov replied:
Done

pkolaczk reviewed (approved):
Looks great! Thank you.

blambov commented:
Some benchmarks for the various variations in the first commit:

Without comparator slowdown:
```
Benchmark                                           (consumeRatio)  (nullChance)  (size)  Mode  Cnt     Score     Error  Units
SortingIteratorBenchmark.testArrayListSortIterator               0           0.1     100  avgt    5     7.867 ±   0.150  us/op
SortingIteratorBenchmark.testArrayListSortIterator               0           0.1    1000  avgt    5    98.926 ±   2.172  us/op
SortingIteratorBenchmark.testArrayListSortIterator               0           0.1   10000  avgt    5  1676.058 ±  25.975  us/op
SortingIteratorBenchmark.testArrayListSortIterator             0.1           0.1     100  avgt    5     7.482 ±   0.086  us/op
SortingIteratorBenchmark.testArrayListSortIterator             0.1           0.1    1000  avgt    5    99.109 ±   3.283  us/op
SortingIteratorBenchmark.testArrayListSortIterator             0.1           0.1   10000  avgt    5  1649.387 ±  40.857  us/op
SortingIteratorBenchmark.testArrayListSortIterator               1           0.1     100  avgt    5     7.795 ±   0.100  us/op
SortingIteratorBenchmark.testArrayListSortIterator               1           0.1    1000  avgt    5   101.922 ±   3.515  us/op
SortingIteratorBenchmark.testArrayListSortIterator               1           0.1   10000  avgt    5  1691.093 ±  44.135  us/op
SortingIteratorBenchmark.testArraySortIterator                   0           0.1     100  avgt    5     7.410 ±   0.040  us/op
SortingIteratorBenchmark.testArraySortIterator                   0           0.1    1000  avgt    5   103.759 ±  19.991  us/op
SortingIteratorBenchmark.testArraySortIterator                   0           0.1   10000  avgt    5  1730.396 ±  29.023  us/op
SortingIteratorBenchmark.testArraySortIterator                 0.1           0.1     100  avgt    5     7.505 ±   0.062  us/op
SortingIteratorBenchmark.testArraySortIterator                 0.1           0.1    1000  avgt    5   103.394 ±   9.151  us/op
SortingIteratorBenchmark.testArraySortIterator                 0.1           0.1   10000  avgt    5  1738.155 ±  86.688  us/op
SortingIteratorBenchmark.testArraySortIterator                   1           0.1     100  avgt    5     7.686 ±   0.173  us/op
SortingIteratorBenchmark.testArraySortIterator                   1           0.1    1000  avgt    5   101.690 ±   0.576  us/op
SortingIteratorBenchmark.testArraySortIterator                   1           0.1   10000  avgt    5  1752.275 ±  42.908  us/op
SortingIteratorBenchmark.testLuceneIterator                      0           0.1     100  avgt    5     2.021 ±   0.323  us/op
SortingIteratorBenchmark.testLuceneIterator                      0           0.1    1000  avgt    5    18.792 ±   0.278  us/op
SortingIteratorBenchmark.testLuceneIterator                      0           0.1   10000  avgt    5   198.632 ±  18.581  us/op
SortingIteratorBenchmark.testLuceneIterator                    0.1           0.1     100  avgt    5     2.585 ±   0.055  us/op
SortingIteratorBenchmark.testLuceneIterator                    0.1           0.1    1000  avgt    5    28.186 ±   1.003  us/op
SortingIteratorBenchmark.testLuceneIterator                    0.1           0.1   10000  avgt    5   347.825 ±   2.267  us/op
SortingIteratorBenchmark.testLuceneIterator                      1           0.1     100  avgt    5     7.088 ±   0.103  us/op
SortingIteratorBenchmark.testLuceneIterator                      1           0.1    1000  avgt    5    99.186 ±   1.558  us/op
SortingIteratorBenchmark.testLuceneIterator                      1           0.1   10000  avgt    5  1474.128 ± 100.739  us/op
SortingIteratorBenchmark.testPriorityQueueIterator               0           0.1     100  avgt    5     3.760 ±   0.036  us/op
SortingIteratorBenchmark.testPriorityQueueIterator               0           0.1    1000  avgt    5    37.578 ±   0.424  us/op
SortingIteratorBenchmark.testPriorityQueueIterator               0           0.1   10000  avgt    5   605.327 ±  88.198  us/op
SortingIteratorBenchmark.testPriorityQueueIterator             0.1           0.1     100  avgt    5     4.440 ±   0.037  us/op
SortingIteratorBenchmark.testPriorityQueueIterator             0.1           0.1    1000  avgt    5    45.539 ±   1.151  us/op
SortingIteratorBenchmark.testPriorityQueueIterator             0.1           0.1   10000  avgt    5   741.981 ± 117.230  us/op
SortingIteratorBenchmark.testPriorityQueueIterator               1           0.1     100  avgt    5     8.170 ±   0.018  us/op
SortingIteratorBenchmark.testPriorityQueueIterator               1           0.1    1000  avgt    5   103.343 ±   2.243  us/op
SortingIteratorBenchmark.testPriorityQueueIterator               1           0.1   10000  avgt    5  1755.651 ±  36.646  us/op
SortingIteratorBenchmark.testSortingIterator                     0           0.1     100  avgt    5     1.789 ±   0.023  us/op
SortingIteratorBenchmark.testSortingIterator                     0           0.1    1000  avgt    5    17.157 ±   0.222  us/op
SortingIteratorBenchmark.testSortingIterator                     0           0.1   10000  avgt    5   195.905 ±   2.036  us/op
SortingIteratorBenchmark.testSortingIterator                   0.1           0.1     100  avgt    5     2.469 ±   0.029  us/op
SortingIteratorBenchmark.testSortingIterator                   0.1           0.1    1000  avgt    5    28.317 ±   1.107  us/op
SortingIteratorBenchmark.testSortingIterator                   0.1           0.1   10000  avgt    5   355.700 ±   4.252  us/op
SortingIteratorBenchmark.testSortingIterator                     1           0.1     100  avgt    5     6.439 ±   0.035  us/op
SortingIteratorBenchmark.testSortingIterator                     1           0.1    1000  avgt    5    98.307 ±   0.793  us/op
SortingIteratorBenchmark.testSortingIterator                     1           0.1   10000  avgt    5  1445.137 ±  26.843  us/op
SortingIteratorBenchmark.testSortingIteratorNulls                0           0.1   10000  avgt    5   217.075 ±  2.047  us/op
SortingIteratorBenchmark.testSortingIteratorNulls              0.1           0.1   10000  avgt    5   359.950 ±  2.929  us/op
SortingIteratorBenchmark.testSortingIteratorNulls                1           0.1   10000  avgt    5  1477.807 ± 11.146  us/op
SortingIteratorBenchmark.testStreamIterator                      0           0.1     100  avgt    5     7.070 ±   0.083  us/op
SortingIteratorBenchmark.testStreamIterator                      0           0.1    1000  avgt    5    93.735 ±   1.143  us/op
SortingIteratorBenchmark.testStreamIterator                      0           0.1   10000  avgt    5  1236.863 ± 242.851  us/op
SortingIteratorBenchmark.testStreamIterator                    0.1           0.1     100  avgt    5     6.961 ±   0.113  us/op
SortingIteratorBenchmark.testStreamIterator                    0.1           0.1    1000  avgt    5    94.358 ±   1.881  us/op
SortingIteratorBenchmark.testStreamIterator                    0.1           0.1   10000  avgt    5  1233.850 ± 180.382  us/op
SortingIteratorBenchmark.testStreamIterator                      1           0.1     100  avgt    5     8.200 ±   0.346  us/op
SortingIteratorBenchmark.testStreamIterator                      1           0.1    1000  avgt    5   103.838 ±   2.760  us/op
SortingIteratorBenchmark.testStreamIterator                      1           0.1   10000  avgt    5  1291.791 ±  19.587  us/op
```
with slowed-down comparator, to reflect the number of comparisons done in each case:
```
Benchmark                                           (comparatorSlowDown)  (consumeRatio)  (nullChance)  (size)  Mode  Cnt      Score     Error  Units
SortingIteratorBenchmark.testArrayListSortIterator                  1000               0           0.1    1000  avgt    5  12256.800 ± 127.062  us/op
SortingIteratorBenchmark.testArrayListSortIterator                  1000             0.1           0.1    1000  avgt    5  12316.817 ±  80.742  us/op
SortingIteratorBenchmark.testArrayListSortIterator                  1000               1           0.1    1000  avgt    5  12408.484 ± 638.007  us/op
SortingIteratorBenchmark.testArraySortIterator                      1000               0           0.1    1000  avgt    5  12341.548 ±  37.559  us/op
SortingIteratorBenchmark.testArraySortIterator                      1000             0.1           0.1    1000  avgt    5  12319.795 ±  59.302  us/op
SortingIteratorBenchmark.testArraySortIterator                      1000               1           0.1    1000  avgt    5  12278.161 ±  78.878  us/op
SortingIteratorBenchmark.testLuceneIterator                         1000               0           0.1    1000  avgt    5   2668.925 ±  29.519  us/op
SortingIteratorBenchmark.testLuceneIterator                         1000             0.1           0.1    1000  avgt    5   5138.790 ±  45.101  us/op
SortingIteratorBenchmark.testLuceneIterator                         1000               1           0.1    1000  avgt    5  18374.428 ± 262.719  us/op
SortingIteratorBenchmark.testPriorityQueueIterator                  1000               0           0.1    1000  avgt    5   3272.795 ±  40.834  us/op
SortingIteratorBenchmark.testPriorityQueueIterator                  1000             0.1           0.1    1000  avgt    5   6065.632 ± 165.797  us/op
SortingIteratorBenchmark.testPriorityQueueIterator                  1000               1           0.1    1000  avgt    5  24220.381 ± 161.914  us/op
SortingIteratorBenchmark.testSortingIterator                        1000               0           0.1    1000  avgt    5   2777.817 ± 376.211  us/op
SortingIteratorBenchmark.testSortingIterator                        1000             0.1           0.1    1000  avgt    5   5457.167 ±  29.412  us/op
SortingIteratorBenchmark.testSortingIterator                        1000               1           0.1    1000  avgt    5  23717.365 ± 255.448  us/op
SortingIteratorBenchmark.testSortingIteratorNulls                   1000               0           0.1    1000  avgt    5   2595.572 ±   7.496  us/op
SortingIteratorBenchmark.testSortingIteratorNulls                   1000             0.1           0.1    1000  avgt    5   3918.546 ±  54.216  us/op
SortingIteratorBenchmark.testSortingIteratorNulls                   1000               1           0.1    1000  avgt    5  12519.753 ± 143.698  us/op
SortingIteratorBenchmark.testStreamIterator                         1000               0           0.1    1000  avgt    5  12273.994 ±  51.229  us/op
SortingIteratorBenchmark.testStreamIterator                         1000             0.1           0.1    1000  avgt    5  12345.579 ± 357.668  us/op
SortingIteratorBenchmark.testStreamIterator                         1000               1           0.1    1000  avgt    5  12305.154 ± 100.336  us/op

```

blambov commented on src/java/org/apache/cassandra/utils/SortingIteratorNulls.java:
```diff
@@ -0,0 +1,144 @@
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+
+public class SortingIteratorNulls<T> implements Iterator<T>
```
There are a couple more differences. After I saw an expected advantage for `LucenePriorityQueue` with slowed-down comparator, I made the second version to test the difference between two approaches:
- get rid of nulls, keep track of an active queue size, replace top item with another live item on removal;
- keep nulls in the list but compare them as bigger than live items, keep the size of the queue constant and replace top item with null on removal.

As described in the comment with the next commit, the latter turned out to be roughly as good (after some further optimizations) with fast comparators, and significantly better with slow ones, mostly because pushing down a null requires half the comparisons.

blambov replied:
The next commit selects `SortingIteratorNulls` as the one to go with, and removes the other implementations.

michaeljmarshall commented on src/java/org/apache/cassandra/utils/TopKSelector.java:
```diff
@@ -0,0 +1,160 @@
+
+    public List<T> getShared()
+    {
+        int size = prepareAndReturnSize();
+        heapSortUpTo(size);
```
It is a pretty minor optimization, but in the TopKProcessor, we don't actually need the results in order, we only need the top k items. Is it worth adding a method to get the top results without also ordering them?

michaeljmarshall replied:
A similar design would be the iterator from java's PriorityQueue implementation. That iterates of the internal array without any guarantees about order.

blambov replied:
This is already available in the `getUnsortedShared` method. Changed TopKProcessor to use it.

michaeljmarshall replied:
Great, thanks

michaeljmarshall reviewed (commented):
Looks great! I left some minor comments.

blambov commented on src/java/org/apache/cassandra/index/sai/utils/RowIdWithScore.java:
```diff
@@ -18,15 +18,12 @@
  * Represents a row id with a score.
- * Note: this class has a natural ordering that is inconsistent with equals in order to
- * use {@link PriorityQueue}'s O(N) constructor.
+ * Note: this class has a natural ordering that is inconsistent with equals.
```
Added a static `compare` method to make them comparable but keep the encapsulation.

Any other place the comparator should be removed?

michaeljmarshall replied:
Thanks, that looks great. The only remaining objects that have this note are `PrimaryKeyWithSortKey` and its subclasses, and those need to be comparable, so we don't have anything left to change.

blambov commented on src/java/org/apache/cassandra/utils/BinaryHeap.java:
```diff
@@ -0,0 +1,309 @@
+    private <T, K extends T> void advanceTo(K targetKey, int heapIndex, BiFunction<T, K, T> itemAdvancer)
+    {
+        if (heapIndex >= heap.length)
+            return;
```
No longer relevant.

blambov commented on src/java/org/apache/cassandra/utils/TopKSelector.java:
```diff
@@ -0,0 +1,160 @@
+        return new ArrayList<>(getShared());
+    }
+
+    public <R> List<R> getTransformed(Function<T, R> transformer)
```
Added a test and kept it.

blambov commented on src/java/org/apache/cassandra/utils/TopKSelector.java:
```diff
@@ -0,0 +1,160 @@
+
+    public List<T> get()
+    {
+        return new ArrayList<>(getShared());
```
I was maybe overly defensive here, but we shouldn't share arrays when they could be scrambled unless the call is specifically marked as a sharing version. Added comments to clarify and added a test for reuse.

Did not change `SortedRowsBuilder` as I believe that class benefits from dropping the `RowWithIdAndScore` encapsulations.

michaeljmarshall replied:
Works for me, thanks

blambov commented:
I am done tinkering with this. It's ready for final review.

michaeljmarshall commented on src/java/org/apache/cassandra/index/sai/disk/vector/BruteForceRowIdIterator.java:
```diff
@@ -61,59 +66,87 @@ public RowWithApproximateScore(int rowId, int ordinal, float appoximateScore)
-                RowWithApproximateScore rowOrdinalScore = approximateScoreQueue.poll();
-                float score = reranker.similarityTo(rowOrdinalScore.ordinal);
-                exactScoreQueue.add(new RowIdWithScore(rowOrdinalScore.rowId, score));
+    public boolean hasNext()
+    {
+        if (next == null)
+        {
+            // Prepare/advance only when the next value is requested.
+            if (!initialized)
+            {
+                for (int i = 0; i < heap.length; ++i)
+                    heap[i] = getNextApproximateAndRerank();
+                heapify();
+                initialized = true;
+            }
+            else
+            {
+                replaceTop(getNextApproximateAndRerank());
```
This changes the logic a bit. The original design drains the `limit` and then refills up to the `topk` count from the remaining approximate values, while the new design eagerly backfills rows. The assumed benefit of the original implementation is that we only need to read additional vectors from storage if there are shadowed primary keys.

I don't have performance data on this though. Do you feel strongly about making it eagerly back fill `RowIdWithScore` entries?

michaeljmarshall replied:
Great, thanks

michaeljmarshall reviewed (approved):
LTGM. Left one question about the `BruteForceRowIdIterator` changes. It doesn't need to block the PR.

blambov commented:
Rebased and squashed in preparation for review. Pre-squash version can be found [here](https://github.com/blambov/cassandra/tree/CNDB-10463-pre-squash).

blambov commented:
@pkolaczk, do you want to take another look, or is this ready for merging?

blambov commented:
The changes and the motivation behind them are also explained in [this presentation](https://docs.google.com/presentation/d/1SbCareOn2Da3J5MK9146eEw6mP55DK-nfMaFOyofmFs/edit?usp=sharing).

blambov commented:
Rebased to solve conflicts, merging if tests are okay.

blambov commented:
The failing tests appear to be flakes, passing locally. Merging.
